src/garage/torch/algos/sac.py

```diff
@@ -0,0 +1,322 @@
+                         min_buffer_size=min_buffer_size,
+                         replay_buffer=replay_buffer,
+                         use_target=True,
+                         discount=discount,
+                         smooth_return=smooth_return)
+        self.reward_scale = reward_scale
+        # use 2 target q networks
+        self.target_qf1 = copy.deepcopy(self.qf1)
+        self.target_qf2 = copy.deepcopy(self.qf2)
+        self.policy_optimizer = optimizer(self.policy.parameters(),
+                                          lr=self.policy_lr)
+        self.qf1_optimizer = optimizer(self.qf1.parameters(), lr=self.qf_lr)
+        self.qf2_optimizer = optimizer(self.qf2.parameters(), lr=self.qf_lr)
+        # automatic entropy coefficient tuning
+        self.use_automatic_entropy_tuning = use_automatic_entropy_tuning
+        if self.use_automatic_entropy_tuning and not alpha:
+            if target_entropy:
+                self.target_entropy = target_entropy
+            else:
+                self.target_entropy = -np.prod(
+                    self.env_spec.action_space.shape).item()
+            self.log_alpha = torch.tensor([self.initial_log_entropy], dtype=torch.float, requires_grad=True)
+            self.alpha_optimizer = optimizer([self.log_alpha], lr=self.policy_lr)
+        else:
+            self.alpha = alpha
+
+        self.episode_rewards = deque(maxlen=30)
+        self.success_history = []
+
+    def train_timing(self, runner):
+        from line_profiler import LineProfiler
+        lp = LineProfiler()
+        lp_wrapper = lp(self._train)
+        lp_wrapper(runner)
+        lp.print_stats()
+        import ipdb; ipdb.set_trace()
+
+    def train_helper(self, runner, batch_size):
+        eval_steps = runner.obtain_samples(-1, 30, add_transitions_to_buffer=False)
+        runner.step_path = runner.obtain_samples(runner.step_itr, batch_size)
+        tabular.record("buffer_size", self.replay_buffer.n_transitions_stored)
+        last_return = self.train_once(runner.step_itr,
+                                        eval_steps)
+        runner.step_itr += 1
+        return last_return
+
+    def _train(self, runner):
+        """Obtain samplers and start actual training for each epoch.
+
+        Args:
+            runner (LocalRunner): LocalRunner is passed to give algorithm
+                the access to runner.step_epochs(), which provides services
+                such as snapshotting and sampler control.
+
+        Returns:
+            The average return in last epoch cycle.
+
+        """
+        last_return = None
+
+        for epoch in runner.step_epochs():
+            if self.replay_buffer.n_transitions_stored < self.min_buffer_size:
+                batch_size = self.min_buffer_size
+            else:
+                batch_size = None
+            for cycle in range(self.n_epoch_cycles):
+                # eval_steps = runner.obtain_samples(-1, 30, add_transitions_to_buffer=False)
```

Comment:
Definitely remove commented code before committing it. If you need it later, put it in another branch.